### PR TITLE
add chained credential provider

### PR DESCRIPTION
**Pull request by balamurugana**

### CODE REVIEW

File: api/src/main/java/io/minio/credentials/ChainProvider.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+package io.minio.credentials;
+
+import java.util.Arrays;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+/** Chained credential provider work with list of credential providers. */
+public class ChainProvider implements Provider {
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/ChainProvider.java:24]: `ChainedProvider` might sound better

File: api/src/main/java/io/minio/credentials/ChainProvider.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        credentials = currentProvider.fetch();
+      } catch (IllegalStateException e) {
+        // Ignore and fallback to iteration.
+      }
+    }
+
+    if (credentials != null) {
+      synchronized (this) {
+        this.credentials = credentials;
+      }
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/ChainProvider.java:52]: Not sure if synchronizing this block helps. Making the `fetch` method itself synchronized might be better.

File: api/src/main/java/io/minio/credentials/ChainProvider.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+      return this.credentials;
+    }
+
+    for (Provider provider : providers) {
+      try {
+        credentials = provider.fetch();
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/ChainProvider.java:59]: Might want to skip doing this for `currentProvider` as that has already happened in previous lines.

File: api/src/main/java/io/minio/credentials/ChainProvider.java (hunk 1 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/credentials/ChainProvider.java:24]: done

File: api/src/main/java/io/minio/credentials/ChainProvider.java (hunk 2 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/credentials/ChainProvider.java:52]: If entire method is locked in execution, `fetch()` becomes sequential execution, that leads to API execution sequential.

File: api/src/main/java/io/minio/credentials/ChainProvider.java (hunk 3 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/credentials/ChainProvider.java:59]: Its OK to try one more time in a loop.

File: api/src/main/java/io/minio/credentials/ChainProvider.java (hunk 2 above)
Comment (anjalshireesh) [reply api/src/main/java/io/minio/credentials/ChainProvider.java]: Can you please help me understand how the current synchronization works? Maybe with an example use case of concurrent requests.

**Review by sinhaashish — APPROVED** [review]
LGTM

**Pull request merged**